### PR TITLE
QFJ-981 / QFJ-982 Remove settings from `DataDictionary` class

### PR DESCRIPTION
**Pull request by philipwhiuk**
Not much testing on this but here's where we're/I'm thinking of going with it.

### CODE REVIEW

**Comment by chrjohn** [comment]
Did not follow this up through the changed code yet but there are NullPointerExceptions at

    at quickfix.DataDictionary.checkHasValue(DataDictionary.java:683)
	at quickfix.DataDictionary.iterate(DataDictionary.java:555)
	at quickfix.DataDictionary.validate(DataDictionary.java:538)
	at quickfix.Session.next(Session.java:998)

Probably forgotten to pass the default validation settings?!

Thanks for the PR!

**Comment by chrjohn** [comment]
Hi @philipwhiuk , I wanted to add a test for QFJ-981 which is related to this issue. However, I do not seem to be able to push changes to this branch. Have you enabled push access for maintainers on this PR?
Thanks,
Chris.

**Comment by philipwhiuk** [comment]
Yep I allowed edits from maintainers... Not sure why it won't let you push..

**Comment by chrjohn** [comment]
Got a 'permission denied'. Maybe got one of the parameters wrong. Will try again later. 

Am 29. November 2019 15:27:31 MEZ schrieb Philip Whitehouse <notifications@github.com>:

**Comment by chrjohn** [comment]
Nevermind, turned out I was using the wrong branch

**Comment by philipwhiuk** [comment]
I'm wondering whether it's better to rename it to `ValidationSettings`. It's kind of a half way between SessionSettings and DataDictionarySettings.

**Comment by chrjohn** [comment]
> I'm wondering whether it's better to rename the new struct to `ValidationSettings`. It's kind of a half way between `SessionSettings` and `DataDictionarySettings`.

Sounds sensible. `ValidationSettings` is more descriptive.

**Comment by philipwhiuk** [comment]
While trying to test this, I tried to design scenario that was easy to QA. I ran into #247 which I didn't expect.

I'm trying to decide whether my calling code should call `DataDictionary.validate` after constructing a message or whether #247 is valid.

**Comment by chrjohn** [comment]
Hi @philipwhiuk , I was thinking about into which version this fix needs to go. When using semantic versioning (which I'm roughly trying to follow) this needs to go into 3.0 due to public interface changes, right?

**Comment by chrjohn** [comment]
I think I failed at trying to resolve the merge conflict. Will check.

**Comment by philipwhiuk** [comment]
> Hi @philipwhiuk , I was thinking about into which version this fix needs to go. When using semantic versioning (which I'm roughly trying to follow) this needs to go into 3.0 due to public interface changes, right?

Yes, it's probably 3.0

**Comment by the-thing** [comment]
@chrjohn 

QFJ-981/QFJ-982 are actually a duplicate of 2016 issue QFJ-877, but there was no investigation so it can be closed once this is merged.

@philipwhiuk 

You probably need to add null checks against `quickfix.ValidationSettings` inside `quickfix.Message` class (possibly in `quickfix.DataDictionary` as well). The code assumes that data dictionary and validation settings are either both null or or none. However, these are public methods and they can be called with any arguments.

- `quickfix.DataDictionary#validate(quickfix.Message, quickfix.ValidationSettings)`
- `quickfix.Message#fromString(java.lang.String, quickfix.DataDictionary, quickfix.DataDictionary, quickfix.ValidationSettings, boolean, boolean)`

Additionally, these methods were not overloaded. This is a breaking change as I have seen these methods being used in test and FIX simulator code while working for at least two clients. This one might not be a big deal and probably does not need to be fixed, but it will add some work for developers integrating with the newest version.

**Comment by philipwhiuk** [comment]
Yeah it's a breaking change. I think the plan is to release this as 3.0.0

I'm happy to review PRs to add null checks.

We're using this internally and it works fine.

**Comment by the-thing** [comment]
Null checks would be good to have, but your choice. I would not worry to much about it.

File: quickfixj-core/src/main/java/quickfix/Message.java [hunk 1]
```diff
@@ -92,27 +92,26 @@ protected Message(int[] fieldOrder) {
...
     }
 
     public Message(String string, boolean validate) throws InvalidMessage {
         initializeHeader();
-        fromString(string, null, validate, true);
+        fromString(string, null, null, validate, true);
     }
 
-    public Message(String string, DataDictionary dd) throws InvalidMessage {
+    public Message(String string, DataDictionary dd, ValidationSettings dds) throws InvalidMessage {
```
Comment (chrjohn) [inline quickfixj-core/src/main/java/quickfix/Message.java:103]: Would it be sensible to have a constructor which does `new ValidationSettings()` if the dictionary is not NULL and if the settings are not passed? That way one wouldn't have to change their code if they are using the default validation settings.

**Comment by chrjohn** [comment]
Hi @philipwhiuk , I have now worked on this to get the changes into master after five years. ;) I hope it is OK that I opened a separate PR for this (somehow got tangled up in the changes of the last, but your commits will of course be included there: https://github.com/quickfix-j/quickfixj/pull/831

Thank you 👍

**Pull request closed without merging**